### ISSUE TITLE
Gracefully handle k8s resource size limit for large Application CRs

### ISSUE THREAD

**Issue by crenshaw-dev** [argoproj/argo-cd#14486]
# Summary

Intuit had an app fail to sync when it hit ~3k resources managed in a single App. I believe the problem was that it attempted to update the sync status, which contained the status of all 3k resources, and we hit the k8s resource size limit.

We should provide more ways for the user to sacrifice certain features/conveniences to allow the resource to fit within the size limit. Ideas below.

# Motivation

3000 really isn't that big a number.

# Proposal

1. store the status.resources and status.operationState.operation.sync.resources fields (or maybe just the whole status field) as gzip/base64 when the app hits a configurable number of managed resources
2. Offload more resource info to Redis, like we did with health data

**Comment by crenshaw-dev** [comment]
Yep, I like this. One question:

> Check if the new CRD state will exceed the 1.5mb etcd limit

How would you propose to perform that check?

**Comment by leoluz** [comment]
> How would you propose to perform that check?

You have the computed `status` field state that is about to be persisted isn't it? Just check its size to drive the persisting logic.

**Comment by crenshaw-dev** [comment]
> You have the computed status field state that is about to be persisted isn't it?

Not necessarily. Some places, e.g. persisting operation state, calculate only a patch: https://github.com/argoproj/argo-cd/blob/15eeb307eb03191e7581d8e616072de4fd4b20e0/controller/appcontroller.go#L1250

Even if we know the full status field contents, I see a few potential problems:
1) you're missing the sizes of top-level keys `metadata`, `spec`, and `operation`
2) marshaling the status before every write operation be a performance drag

I'd suggest a lightweight, configurable heuristic like "if it manages > N resources, compress."

**Comment by leoluz** [comment]
> I'd suggest a lightweight, configurable heuristic like "if it manages > N resources, compress."

Yes.. I like that too.

**Comment by zswanson** [comment]
Related, we are looking to enable argo with a large scale of applications soon (5k+) and we're concerned about hitting GKE limits where any single resource type in etcd must be < 800MB. An option to *always* compress statuses, regardless of number of resources, would be nice.

**Comment by cjin62** [comment]
Any further updates on when ArgoCD will be able to implement the improvements?

**Comment by thesuperzapper** [comment]
I just want to highlight how important it is for ArgoCD to handle Apps with large numbers of resources.

I like the Idea of compressing `status.resources` as this tends to be the largest part of the Application spec and there is a hard ~1.5Mb limit on any resource.

**Comment by Reversaidx** [comment]
We faced the same issue, do we have any idea what to do?(

**Comment by Lirt** [comment]
Hello,

I'm also affected by this limitation. Is there a development plan to mitigate this? Perhaps the idea with redis sound as the most robust. The idea to put data that scales with amount resources into status of Kubernetes resource sounds a bit non-optimal but I guess not many devs thought users would create applications that synchronize 10k kubernetes resources into the cluster 😄.

I was trying to find at least a workaround, but I wasn't successful. I tried to disable synchronization which didn't help. I think that `status.resources` is used internally in multiple places in the code so it's also not so trivial to perhaps make it a feature toggle..

Anyway, if you had some idea for a workaround, it would help.

**Comment by wsilva** [comment]
We are facing same issues. For now we set the history manually to 3 to mitigate. But if there are any change that can be done to solve in long term please let us know. 🙇 

<img width="946" alt="Image" src="https://github.com/user-attachments/assets/0d80345a-95ee-4330-ad70-fe0c819ed0c5" />

**Comment by ajay0988** [comment]
I am also facing this issue , tried below things but no luck (This is kube-prometheus-stack ,chart version 72.6.2)

1.  revisionHistoryLimit: 3 
2. server.enable.gzip: "true" in argocd-cmd-params-cm configmap.

ArgoCD version : v2.13

**Comment by TotallyTemporary** [comment]
We had this problem as well, ArgoCD kept getting stuck at around 6000 resources and it's sometimes getting difficult to unstuck it.
One observation is that (at least for us), the resources data is very repetitive, with the same apiVersion, kind, namespace appearing for each resource.

**Comment by elisiano** [comment]
We've been bitten by this too (almost creating an outage 😬).

Our team had to refactor helm charts to use sharding, but it makes the whole setup confusing.
Having a built-in solution for this would be much better.

Just like others, I am wondering if there is any plan to address this.

**Comment by nitishfy** [comment]
i'll take this up.

**Comment by crenshaw-dev** [comment]
One easy win here would be to have the Applicaiton controller fall back to _only_ writing an Error condition. Basically deep-copy the old `status` field, set an error condition, and apply that, discarding the status that's failing to apply.

**Comment by crenshaw-dev** [comment]
Trick I just discovered: set `ApplyOutOfSyncOnly=true`. That's probably gonna _way_ reduce what gets stored in `status.operationState.syncResult.resources` which would otherwise be similar size to `status.resources` (actually a bit bigger).
